lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp

```diff
@@ -1288,112 +1292,101 @@ void HdVP2Mesh::_UpdateDrawItem(
+                drawScene.GetSelectionHighlightColor(true)
+            };
+
+            std::vector<unsigned char> colorIndices;
+            colorIndices.resize(instanceCount, 0);
+
+            if (const auto state = drawScene.GetActiveSelectionState(id)) {
+                for (const auto& indexArray : state->instanceIndices) {
                     for (const auto index : indexArray) {
-                        selectedIndices[index] = true;
+                        colorIndices[index] = 1;
                     }
                 }
-                for (unsigned int index=0; index<instanceCount; index++)
-                {
-                    if (!selectedIndices[index])
-                        continue;
-                    transforms[index].Get(instanceMatrix.matrix);
-                    instanceMatrix = worldMatrix * instanceMatrix;
-                    stateToCommit._instanceTransforms.append(instanceMatrix);
-                }
-#else
+            }
+
+            if (const auto state = drawScene.GetLeadSelectionState(id)) {
                 for (const auto& indexArray : state->instanceIndices) {
                     for (const auto index : indexArray) {
-                        transforms[index].Get(instanceMatrix.matrix);
-                        instanceMatrix = worldMatrix * instanceMatrix;
-                        stateToCommit._instanceTransforms.append(instanceMatrix);
+                        colorIndices[index] = 2;
                     }
                 }
-#endif
-            }
-        }
-        else {
-            stateToCommit._instanceTransforms.setLength(instanceCount);
-            for (size_t i = 0; i < instanceCount; ++i) {
-                transforms[i].Get(instanceMatrix.matrix);
-                instanceMatrix = worldMatrix * instanceMatrix;
-                stateToCommit._instanceTransforms[i] = instanceMatrix;
             }
 
-            // If the item is used for both regular draw and selection highlight,
-            // it needs to display both wireframe color and selection highlight
-            // with one color vertex buffer.
-            if (drawItem->ContainsUsage(HdVP2DrawItem::kSelectionHighlight)) {
-                const MColor& color = drawScene.GetWireframeColor();
-                unsigned int offset = 0;
-
-                stateToCommit._instanceColors.setLength(instanceCount * kNumColorChannels);
-                for (unsigned int i = 0; i < instanceCount; ++i) {
-                    for (unsigned int j = 0; j < kNumColorChannels; j++) {
-                        stateToCommit._instanceColors[offset++] = color[j];
-                    }
-                }
+            for (unsigned int i = 0; i < instanceCount; i++) {
+                // If it is a dedicated selection highlight item, we don't draw wireframe for
+                // unselected instances.
+                unsigned char colorIndex = colorIndices[i];
+                if (isDedicatedSelectionHighlightItem && colorIndex == 0)
```

Comment:
We have a goal of sharing more of the code between rprims, e.g. mesh and curves. When looking at the code changes, I see there are differences between the two (e.g. isDedicatedSelectionHighlightItem is removed in one and stayed here...color is appended here instead of more performant resize of the vector and writing at index).

I have two questions:
1. Are there any reasons selection highlighting color support between mesh and curves be different?
2. We were planning to do some refactoring when we will add another rprim support that has similar needs...but looking at changes like this one, maybe we should do it sooner? (not this PR)

Reply:
1. I tried to avoid a dedicated selection highlight item when implementing the curves, but I've already realized that we couldn't do that because we still need it for refineLevel > 0 case and instanced curves (#259 ). So we still need the same selection highlight logic as mesh ultimately, but for now there is no way to test the dedicated selection highlight item for curves, that is why I removed that `isDedicatedSelectionHighlightItem` code branch in basisCurves for now.

2. I think when we add the third rprim support, we will be clearer which code are common and which are not to all rprims, not just selection highlight, but also the repr management code. I don't think we have to do it sooner, the refactoring work itself is a lot heavier than duplicating two copies of code.